Add reusable `reviewing-openmc-code` skill and Copilot Review agent (#3842)
openmc-dev/openmc: .claude/skills/openmc-code-review/SKILL.md

---

Apply repository-wide guidance from `AGENTS.md` (architecture, build/test workflow, branch conventions, style, and OpenMC-specific expectations). This skill adds a structured review process and rubric on top of that baseline knowledge.

## Determine Review Context

1. **Identify what to review.** Determine the review context based on what the user provides or what can be inferred:
   - **Default (no explicit context):** Compare the current branch against `develop` using `git diff develop...HEAD` to see only commits unique to the current branch.
   - **User specifies a base branch or commit range:** Use that instead.
   - **PR context provided by tooling:** Use the base/head refs supplied by the tool.
2. **Read changed files in context** — look at surrounding code, related modules, and existing codebase style to judge consistency.
3. **Determine review mode:**
   - **Self-review mode** (author preparing changes before submission): Emphasize likely reviewer objections, missing tests/docs, API naming consistency, and quick fixes that would speed up review.

>    - **Reviewer mode** (default): Provide a standard structured review suitable for a code review or PR review.

Reviewer: I'm wondering if having the two different modes is useful? It seems like the main reviewer mode should cover everything that the self-review mode would do as well (such that they're redundant), though perhaps you found some utility here? Have you noticed that the quality/quantity of the issues it flags changes significantly between the two modes? I have so far only tested it in review mode, but can do some more testing.

On the one hand it doesn't seem like it would hurt anything, on the other though, I have found that making the prompts more complex does potentially reduce the amount of time it will spend thinking on the actual review.

Author: This is a good point, and along with your point about it looking "authoritative", I'm thinking it's probably best not to have the skill actually recommend approving/rejecting but rather just point out the issues. The human reviewer should be the one making the judgment call based on the information presented. In any case, I'm going to restructure to not have multiple modes since like you said, the work that it's doing should be the same.

   If the user explicitly requests self-review, use self-review mode. Otherwise, default to reviewer mode.

## Review Criteria

Assess each of the following areas, noting any issues found. If an area looks good, briefly confirm it passes.

### Purpose and Scope
- Do the changes have a clear, well-defined purpose?
- Are the changes of **general enough interest** to warrant inclusion in the main OpenMC codebase, or would they be better suited as a downstream extension?

### Correctness and Testing

> - Do the changes compile and appear functionally correct?

Reviewer: Might strengthen language to get more effort out of the LLM potentially.

```suggestion
- Do the changes compile and can you confirm all logic to be functionally correct?
```

- Are new XML input attributes described in the input reference?
- Are any deprecations or breaking changes clearly noted?

### Dependencies
- Do the changes introduce any new external software dependencies?
- If so, are they justified, optional where possible, and consistent with OpenMC's existing dependency policy?

## Output Format

Produce your review as a structured report with the following sections:

**Context**: State the review mode (self-review or reviewer) and what is being compared (e.g., "current branch vs. `develop`", or the specific commit range/PR).

**Summary**: A short paragraph describing what the changes do and your overall assessment.

**Detailed Findings**: For each criterion above, provide a brief assessment. Use `✓` for items that pass and flag issues with severity:
- `[Minor]` — Style nits, small improvements, non-blocking suggestions
- `[Moderate]` — Issues worth addressing but not strictly blocking
- `[Major]` — Problems that should be resolved before merging

Group findings into:
1. **Blocking issues** — Would justify requesting changes before merge
2. **Non-blocking suggestions** — Improvements that could be addressed now or later

> 3. **Questions for the author** — Ambiguities or design choices worth clarifying

Reviewer: I found that it would often ask questions that it was likely able to answer itself. E.g., "Can you confirm this is correct.." etc. The following proposal helps cut down on noise and forces it to think through things more.
```suggestion
3. **Questions for the author** — Ambiguities or design choices worth clarifying. Do not include questions that you are capable of answering yourself via deep analysis of the full repository. Only propose alternative designs if you have coded them up and can confirm they offer tangible improvements.
```

**Recommended Action**:
- *Reviewer mode*: One of — *Approve*, *Approve with minor suggestions*, *Request changes*, or *Reject* — with a brief justification.

> - *Self-review mode*: A prioritized list of pre-submission fixes — missing tests, undocumented parameters, naming inconsistencies, or anything else likely to slow down review.

Reviewer: The report it generates is very official looking and well organized. Personally, I think I sort of unconsciously assume it must be authoritative and probably is capturing nearly everything, even though logically I know the limitations of LLMs and how complex the OpenMC codebase is. Maybe we'll tune this out after a few reviews, but it seems like it could be useful to give a warning here so as not to instill a false sense of confidence in the report. I've taken a crack at that, but you're welcome to axe this or go a different route.

```suggestion
- *Self-review mode*: A prioritized list of pre-submission fixes — missing tests, undocumented parameters, naming inconsistencies, or anything else likely to slow down review.

**Disclaimer**
Empirical evidence from Zeng et al. (https://arxiv.org/html/2509.01494v1) indicates LLMs (as of late 2025) can only find a very small percentage of bugs in PRs. OpenMC developer experience aligns well with the Zeng et al. findings. Thus, while this agent-based analysis is useful, it is likely only spotting a tiny fraction of the issues in complex PRs.
```
